pkg/meta/openfile.go

```diff
@@ -0,0 +1,155 @@
+package meta
+
+import (
+	"sync"
+	"time"
+)
+
+type openFile struct {
+	attr      Attr
+	refs      int
+	lastCheck time.Time
+	chunks    map[uint32][]Slice
+}
+
+type openfiles struct {
+	sync.Mutex
+	expire time.Duration
+	files  map[Ino]*openFile
+}
+
+func newOpenFiles(expire time.Duration) *openfiles {
+	of := &openfiles{
+		expire: expire,
+		files:  make(map[Ino]*openFile),
+	}
+	go of.cleanup()
+	return of
+}
+
+func (o *openfiles) cleanup() {
+	for {
+		o.Lock()
+		cutoff := time.Now().Add(-time.Hour)
+		for ino, of := range o.files {
+			if of.refs <= 0 && of.lastCheck.Before(cutoff) {
+				delete(o.files, ino)
+			}
+		}
+		o.Unlock()
+		time.Sleep(time.Second)
+	}
+}
+
+func (o *openfiles) OpenCheck(ino Ino, attr *Attr) bool {
+	o.Lock()
+	defer o.Unlock()
+	of, ok := o.files[ino]
+	if ok && time.Since(of.lastCheck) < o.expire {
+		if attr != nil {
+			*attr = of.attr
+			attr.KeepCache = true
+		}
+		of.refs++
+		return true
+	}
+	return false
+}
+
+func (o *openfiles) Open(ino Ino, attr *Attr) {
+	o.Lock()
+	defer o.Unlock()
+	of, ok := o.files[ino]
+	if !ok {
+		of = &openFile{}
+		of.chunks = make(map[uint32][]Slice)
+		o.files[ino] = of
+	} else if attr != nil && attr.Mtime == of.attr.Mtime && attr.Mtimensec == of.attr.Mtimensec {
+		attr.KeepCache = true
+	} else {
```

Comment:
attr.KeepCache = false?

Reply:
This is the only place to assign KeepCache (initialized as zero).